From e8be7cdedc41ff28876d66b566770929d03acbb8 Mon Sep 17 00:00:01 2001
From: Stephen Hemminger <stephen@networkplumber.org>
Date: Sat, 9 May 2026 10:03:25 -0700
Subject: [PATCH] net/sched: netem: handle multi-segment skb in corruption

[ Upstream commit e8be7cdedc41ff28876d66b566770929d03acbb8 ]

The packet corruption code only flipped bits in the linear
header portion of the skb, skipping corruption when
skb_headlen() was zero.

Linearize the whole skb if necessary before corruption.
Extends d64cb81dcbd5 ("net/sched: sch_netem: fix out-of-bounds access
in packet corruption") with a more general solution.

Signed-off-by: Stephen Hemminger <stephen@networkplumber.org>
Link: https://patch.msgid.link/20260509171123.307549-5-stephen@networkplumber.org
Signed-off-by: Paolo Abeni <pabeni@redhat.com>
[Ankit: Backported for v6.1.175. v6.1.x uses prandom_u32_max() in place
 of get_random_u32_below().]
Signed-off-by: Ankit Jain <ankit-aj.jain@broadcom.com>
---
 net/sched/sch_netem.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/net/sched/sch_netem.c b/net/sched/sch_netem.c
index 5d6e928ab750..89e13659d19f 100644
--- a/net/sched/sch_netem.c
+++ b/net/sched/sch_netem.c
@@ -501,16 +501,17 @@ static int netem_enqueue(struct sk_buff *skb, struct Qdisc *sch,
 			qdisc_qstats_drop(sch);
 			goto finish_segs;
 		}
-		if (skb->ip_summed == CHECKSUM_PARTIAL &&
-		    skb_checksum_help(skb)) {
+		if (skb_linearize(skb) ||
+		    (skb->ip_summed == CHECKSUM_PARTIAL && skb_checksum_help(skb))) {
 			qdisc_drop(skb, sch, to_free);
 			skb = NULL;
 			goto finish_segs;
 		}
 
-		if (skb_headlen(skb))
-			skb->data[prandom_u32_max(skb_headlen(skb))] ^=
-				1 << prandom_u32_max(8);
+		if (skb->len) {
+			u32 offset = prandom_u32_max(skb->len);
+			skb->data[offset] ^= 1 << prandom_u32_max(8);
+		}
 	}
 
 	if (unlikely(sch->q.qlen >= sch->limit)) {
-- 
2.39.5
